cider/tests/utils/CiderBatchChecker.cpp

```diff
@@ -21,6 +21,367 @@
+  auto valid_bitmap = batch->getNulls();
+
+  if (!CiderBitUtils::isBitSetAt(valid_bitmap, row_index)) {
+    return NULL_VALUE;
+  } else {
+    std::stringstream fps;
+    fps.clear();
+    double value = data_buffer[row_index];
+    fps << std::setprecision(16) << value;
+    return fps.str();
+  }
+}
+
+std::string CiderBatchStringifier::stringifyStructBatchAt(CiderBatch* batch,
+                                                          int row_index) {
+  if (!batch) {
+    CIDER_THROW(CiderRuntimeException, "StructBatch is nullptr.");
+  }
+
+  ConcatenatedRow row;
+  int col_num = batch->getChildrenNum();
+  auto valid_bitmap = batch->getNulls();
+
+  if (valid_bitmap) {
+    if (!CiderBitUtils::isBitSetAt(valid_bitmap, row_index)) {
+      // this usually should not happen, values in struct batch are expected to be valid
+      // but just in case
+      return NULL_VALUE;
+    }
+  }
+
+  for (int col_index = 0; col_index < col_num; col_index++) {
+    auto child = batch->getChildAt(col_index);
+    std::string cell_str;
+
+    switch (child->getCiderType()) {
+      case SQLTypes::kBOOLEAN:
+        CIDER_THROW(CiderCompileException, "Boolean is currently not supported.");
+      case SQLTypes::kTINYINT:
+        cell_str =
+            stringifyScalarBatchAt<int8_t>(child->as<ScalarBatch<int8_t>>(), row_index);
+        break;
+      case SQLTypes::kSMALLINT:
+        cell_str =
+            stringifyScalarBatchAt<int16_t>(child->as<ScalarBatch<int16_t>>(), row_index);
+        break;
+      case SQLTypes::kINT:
+        cell_str =
+            stringifyScalarBatchAt<int32_t>(child->as<ScalarBatch<int32_t>>(), row_index);
+        break;
+      case SQLTypes::kBIGINT:
+        cell_str =
+            stringifyScalarBatchAt<int64_t>(child->as<ScalarBatch<int64_t>>(), row_index);
+        break;
+      case SQLTypes::kFLOAT:
+        cell_str =
+            stringifyScalarBatchAt<float>(child->as<ScalarBatch<float>>(), row_index);
+        break;
+      case SQLTypes::kDOUBLE:
+        cell_str =
+            stringifyScalarBatchAt<double>(child->as<ScalarBatch<double>>(), row_index);
+        break;
+      case SQLTypes::kSTRUCT:
+        cell_str =
+            stringifyStructBatchAt(dynamic_cast<StructBatch*>(child.get()), row_index);
+        break;
+      default:
+        CIDER_THROW(CiderCompileException, "Unsupported type for stringification.");
+    }
+
+    CHECK(cell_str.size());
+    row.addCol(cell_str);
+  }
+
+  row.finish();
+  return row.getString();
+}
+
+std::vector<ConcatenatedRow> CiderBatchChecker::arrowToConcatenatedRowVector(
+    const std::vector<std::shared_ptr<CiderBatch>>& cider_batches) {
+  std::vector<ConcatenatedRow> total_row;
+  for (auto batch : cider_batches) {
+    auto col_num = batch->getChildrenNum();
+
+    for (int row_index = 0; row_index < batch->getLength(); row_index++) {
+      auto batch_str =
+          CiderBatchStringifier::stringifyStructBatchAt(batch.get(), row_index);
```

Comment:
I guess performance here will be bad, maybe we can construct typed Stringfier before loop